File: electron/src/js/pages/statistics.js
import { api } from "../api.js";
import { el, formatRuntime } from "../ui.js";
import { factBox } from "../components.js";

export default async function render(container) {
  const data = await api.get("/api/statistics");
  const overview = data.overview || {};

  container.append(el("div", { class: "page-head" },
    el("div", {}, el("h1", {}, "Statistics"),
      el("div", { class: "sub" }, "Your library at a glance"))));

  const cards = [
    ["Movies", overview.movies],
    ["TV shows", overview.tv_shows],
    ["Episodes", overview.episodes],
    ["Artists", overview.artists],
    ["Albums", overview.albums],
    ["Tracks", overview.tracks],
    ["People", overview.people],
    ["Files", overview.files],
    ["Watch time", formatRuntime(overview.watch_seconds || 0)],
    ["Watched movies", overview.watched_movies],
    ["Watched episodes", overview.watched_episodes],
    ["Favorites", overview.favorite_count],
    ["Watchlist", overview.watchlist_count],
    ["Sessions", overview.playback_sessions],
    ["Missing files", overview.missing_files],
    ["Artwork cached", overview.artwork_cached],
  ];
  const grid = el("div", { class: "stat-grid" });
  for (const [label, value] of cards) grid.append(factBox(label, value ?? "—"));
  container.append(grid);

  const barsSection = (title, rows, unitFormatter = (v) => v) => {
    if (!rows || !rows.length) return null;
    const max = Math.max(...rows.map((row) => Number(row[1]) || 0), 1);
    const section = el("div", { class: "section" }, el("div", { class: "section-head" }, el("h2", {}, title)));
    const bars = el("div", { class: "bars" });
    for (const [label, value] of rows) {
      bars.append(el("div", { class: "bar-row" },
        el("span", {}, String(label)),
        el("div", { class: "bar" }, el("span", { style: { width: `${((Number(value) || 0) / max) * 100}%` } })),
        el("span", { style: { textAlign: "right", color: "var(--text-dim)" } }, String(unitFormatter(value)))));
    }
    section.append(bars);
    return section;
  };

  container.append(
    barsSection("Top genres", data.top_genres) || el("div"),
    barsSection("Most seen actors", data.top_actors) || el("div"),
    barsSection("Top directors", data.top_directors) || el("div"),
    barsSection("Watch time by month", data.watch_time_by_month, (value) => formatRuntime(value)) || el("div"));
}

File: ui/screens/statistics.py
"""Statistics screen: real numbers from the database."""
from __future__ import annotations

from PyQt6.QtWidgets import (
    QGridLayout,
    QHBoxLayout,
    QLabel,
    QScrollArea,
    QVBoxLayout,
    QWidget,
)

from ui.components.detail_body import _section_title
from ui.screens.base import Screen


class StatCard(QWidget):
    def __init__(self, value: str, label: str, parent=None) -> None:
        super().__init__(parent)
        self.setObjectName("StatCard")
        layout = QVBoxLayout(self)
        value_label = QLabel(value)
        value_label.setObjectName("BigValue")
        layout.addWidget(value_label)
        name_label = QLabel(label)
        name_label.setObjectName("MutedLabel")
        layout.addWidget(name_label)


class BarList(QWidget):
    def __init__(self, entries: list[tuple[str, int]], parent=None) -> None:
        super().__init__(parent)
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(4)
        max_value = max((value for _name, value in entries), default=1) or 1
        for name, value in entries:
            row = QHBoxLayout()
            row.setContentsMargins(0, 0, 0, 0)
            name_label = QLabel(str(name))
            name_label.setFixedWidth(220)
            row.addWidget(name_label)
            bar = QLabel("█" * max(1, int(value / max_value * 34)))
            bar.setObjectName("MutedLabel")
            row.addWidget(bar, 1)
            count = QLabel(str(value))
            count.setObjectName("MutedLabel")
            row.addWidget(count)
            holder = QWidget()
            holder.setLayout(row)
            layout.addWidget(holder)
        layout.addStretch(1)


def format_hours(seconds: float) -> str:
    hours = seconds / 3600
    if hours >= 100:
        return f"{hours:.0f} h"
    if hours >= 1:
        return f"{hours:.1f} h"
    return f"{int(seconds // 60)} min"


class StatisticsScreen(Screen):
    title = "Statistics"

    def __init__(self, context, parent=None) -> None:
        super().__init__(context, parent)
        self.header = QLabel("Statistics")
        self.header.setObjectName("ScreenTitle")
        self.root.addWidget(self.header)
        self.scroll = QScrollArea()
        self.scroll.setWidgetResizable(True)
        self.scroll.setFrameShape(QScrollArea.Shape.NoFrame)
        self.root.addWidget(self.scroll, 1)
        self.content = QWidget()
        self.content_layout = QVBoxLayout(self.content)
        self.content_layout.setSpacing(18)
        self.scroll.setWidget(self.content)

    def refresh(self) -> None:
        from ui.app.context import run_async

        def work():
            svc = self.context.services
            profile = self.context.profile_id
            return {
                "overview": svc.statistics.overview(profile),
                "genres": svc.statistics.top_genres(10),
                "actors": svc.statistics.top_people("actor", 10),
                "directors": svc.statistics.top_people("director", 10),
                "months": svc.statistics.watch_time_by_month(profile, 12),
            }

        run_async(work, self._populate)

    def _populate(self, data: dict) -> None:
        while self.content_layout.count():
            item = self.content_layout.takeAt(0)
            if item.widget():
                item.widget().deleteLater()
        overview = data["overview"]

        grid = QGridLayout()
        grid.setSpacing(10)
        cards = [
            (str(overview["movies"]), "Movies"),
            (str(overview["tv_shows"]), "TV shows"),
            (str(overview["seasons"]), "Seasons"),
            (str(overview["episodes"]), "Episodes"),
            (str(overview["watched_movies"]), "Movies watched"),
            (str(overview["watched_episodes"]), "Episodes watched"),
            (format_hours(overview["watch_seconds"]), "Watch time"),
            (str(overview["artists"]), "Artists"),
            (str(overview["albums"]), "Albums"),
            (str(overview["tracks"]), "Tracks"),
            (str(overview["favorite_count"]), "Favorites"),
            (str(overview["watchlist_count"]), "Watchlist"),
            (str(overview["people"]), "People indexed"),
            (str(overview["files"]), "Media files"),
            (str(overview["artwork_cached"]), "Artwork cached"),
            (str(overview["missing_files"]), "Missing files"),
        ]
        for index, (value, label) in enumerate(cards):
            grid.addWidget(StatCard(value, label), index // 4, index % 4)
        holder = QWidget()
        holder.setLayout(grid)
        self.content_layout.addWidget(holder)

        if data["genres"]:
            self.content_layout.addWidget(_section_title("Top genres"))
            self.content_layout.addWidget(BarList(data["genres"]))
        if data["actors"]:
            self.content_layout.addWidget(_section_title("Top actors in library"))
            self.content_layout.addWidget(BarList(data["actors"]))
        if data["directors"]:
            self.content_layout.addWidget(_section_title("Top directors in library"))
            self.content_layout.addWidget(BarList(data["directors"]))
        if data["months"]:
            self.content_layout.addWidget(_section_title("Watch time by month (hours)"))
            self.content_layout.addWidget(
                BarList([(month, round(hours / 3600, 1)) for month, hours in data["months"]])
            )
        self.content_layout.addStretch(1)
